### ISSUE TITLE
EMF images aren't supported

### ISSUE THREAD

**Issue by mwilliamson** [mwilliamson/mammoth.js#36]
Originally raised on the WordPress plugin: https://wordpress.org/support/topic/some-images-dont-uploadimport-correctly

Adding EMF support in JavaScript may be tricky, but at the very least we can spit out a message explaining why the image won't display properly.

**Comment by mwilliamson** [comment]
A warning is now emitted when image types are detected that a web browser probably won't be able to render properly, including `image/x-emf`.

**Comment by occasl** [comment]
Would there be any possibility of adding EMF/WMF support by using something like PHP's [Imagick](http://php.net/manual/en/book.imagick.php)? Of course it adds a dependency, but this is a real show stopper for us since our legacy docs have these images embedded in them.

**Comment by dividor** [comment]
Just to add, we post-process the Python-mammoth generated HTML to convert these images to PNG, exactly as you mention, using Imagick.

**Comment by occasl** [comment]
If you can add this to branch of this code base, I'm sure you'd make a lot of people happy. :)

**Comment by dividor** [comment]
@occasl I'll try and fork, but in the meantime in case this is useful for anybody ...

I have an image handler to write image files to folder, renaming x-wmf ,x-emf to just be wmf.

`
# Image handler
class ImageWriter(object):

    def __init__(self, output_dir):
        self._output_dir = output_dir
        self._image_number = 1
        if not os.path.isdir(output_dir):
            os.mkdir(output_dir)

    def __call__(self, element):
        extension = element.content_type.partition("/")[2]
        # We rename these so they get processed, turns out they work the same
        if extension in ['x-wmf','x-emf']:
            extension = 'wmf'
        image_filename = "image" + "{0}.{1}".format(self._image_number, extension)
        with open(os.path.join(self._output_dir, image_filename), "wb") as image_dest:
            with element.open() as image_source:
                shutil.copyfileobj(image_source, image_dest)

        self._image_number += 1
        return {"src": image_filename}`

This is called from Mammoth ...

'convert_image = mammoth.images.inline(ImageWriter("./output/" + safe_file_name + '/media'))
        result = mammoth.convert_to_html(docx_file,convert_image=convert_image,style_map=style_map,
                                                   transform_document = mammoth.transforms.run(transform_run))'

The for the image conversion to PNG using Imagemagik, my code calls these commands for any .wmf or .emf files ...

`unoconv -f png -o file.png <file.[emf|wmf]>
convert file.png -flatten -fuzz 1% -trim +repage file.png`

Finally, I update the created html referencing .wmf/.emf to reference .png instead using 

'html_string = result.value
html_string = re.sub("\.wmf", ".png", html_string)
html_string = re.sub("\.emf", ".png", html_string)`

**Comment by mwilliamson** [comment]
You could do all the work in an image converter. That is, if you detect an EMF, pass it to unoconv, and write the result directly to the output file.

**Comment by mwilliamson** [comment]
I've added something along those lines to the Python implementation on the master branch. Use it like so:

```python

def compose(f, g):
    def composed(*args, **kwargs):
        return f(g(*args, **kwargs))
    
    return composed
    

fileobj = open("document.docx", "rb")
result = mammoth.convert_to_html(
    fileobj,
    convert_image=compose(
        mammoth.images.data_uri,
        mammoth.images.libreoffice_wmf_conversion(post_process=mammoth.images.imagemagick_trim),
    ),
)
```

It seems to work when I tried it out.

**Comment by dividor** [comment]
Ah, now that's a much more elegant way of doing it! Will try it out this week. Thanks!

**Comment by BahyQorrect** [comment]
@mwilliamson Thanks for mention. is there any solution for this conversion you did in python to be in Angular or JS?

**Comment by mwilliamson** [comment]
I don't see a reason why the above couldn't be ported to run on node.js, but running in the browser won't work since you can't spawn child processes.

**Comment by BahyQorrect** [comment]
@mwilliamson Yes I want it to work in the browser, as i want the graph to be shown in my system. it appears like broken image so should it be converted to another type before entering the mammoth function to convert all to html?

**Comment by swxxii** [comment]
Thanks @mwilliamson for the python example. I am having the x-emf issue as well. But I tried running it and get an error Sorry if this is a n00b question.
```Traceback (most recent call last):
  File "/Users/simon/TEMP/convert.py", line 15, in <module>
    mammoth.images.libreoffice_wmf_conversion(post_process=mammoth.images.imagemagick_trim),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: module 'mammoth.images' has no attribute 'libreoffice_wmf_conversion'
```
So I'm just wondering if I'm getting the right version of mammoth. I think I installed [1.6.0](https://pypi.org/project/mammoth/) using pip. I am using macOS Sonoma.
